app/services/event_log/journey_runner.rb

```diff
@@ -0,0 +1,33 @@
+module EventLog
+  class JourneyRunner
+    attr_reader :journey
+
+    def initialize(journey)
+      @journey = journey
+    end
+
+    # Process events in order of client_timestamp
+    def call
+      # iterate over all events in the log and apply changes to the journey
+      events.each do |event| # NB: do not use events.find_each as it will break the ordering
+        case event.event_name
+        when Event::CANCEL, Event::COMPLETE, Event::REJECT, Event::START, Event::UNCANCEL, Event::UNCOMPLETE
+          # with a state-changing event, delegate it to the state machine
+          journey.send(event.event_name)
+        when Event::LOCKOUT, Event::LODGING
+          # no action to perform when a journey is locked out or lodged, these events are purely for auditing
+          # TODO: handle other journey events here
+        end
+      end
+
+      # save the journey if it has changed
+      journey.save! if journey.changed? && journey.valid?
```

Comment:
re moving the `save!` to line 16 (inside the loop) - I don't think that's a good idea as it would be called multiple times whilst iterating over the journey's events.

Good point about the validity checker. The problem is that journey events could be temporarily invalid (as they are received in a random/delayed order). And I don't want to reject events even if they are temporarily invalid... So I think I'll change it to:
`journey.save if journey.changed? && journey.valid?`